### DISCUSSION TITLE
TemplatedControl property bound to property in another TemplatedControl does not update

### DISCUSSION THREAD

**Discussion by AdamIzon** [AvaloniaUI/Avalonia#17685] [Q&A]
I have this TemplatedControl, ProjectItemTreeControl, that has a TabManager property that is bound to an object in my View Model
```
public class ProjectItemTreeControl : TemplatedControl
{
    ...

    public static readonly StyledProperty<TabManagerSingleton> TabManagerProperty
        = AvaloniaProperty.Register<ProjectItemTreeControl, TabManagerSingleton>(nameof(TabManager));

    public TabManagerSingleton TabManager // Just gets passed on to DisplayInWidgetControl
    { 
        get { return GetValue(TabManagerProperty); }
        set { SetValue(TabManagerProperty, value); }
    }

    ...
}
```
This TabManager Property in DisplayInWidgetControl gets bound to the TabManager property in DisplayInWidgetControl via TemplateBinding
```
public class DisplayInWidgetControl : TemplatedControl
{
    ...

    public static readonly StyledProperty<TabManagerSingleton> TabManagerProperty 
        = AvaloniaProperty.Register<DisplayInWidgetControl, TabManagerSingleton>(nameof(TabManager));

    public TabManagerSingleton TabManager {
        get { return GetValue(TabManagerProperty); }
        set { SetValue(TabManagerProperty, value); }
    }

    ...
}
```

```
	<ControlTheme x:Key="{x:Type controls:ProjectItemTreeControl}" TargetType="controls:ProjectItemTreeControl">
		<Setter Property="Template">
			<ControlTemplate>
				<TreeView Name="tree" ItemsSource="{TemplateBinding ProjectItems}">
					<TreeView.ItemTemplate>
						<TreeDataTemplate ItemsSource="{Binding ChildrenVisual}" x:DataType="vm:IProjectItemViewModel">
							<Panel>
								<controls:ProjectItemControl 
									Name="projectItemControl" 
									IconPath="{Binding IconPath}" 
									ItemName="{Binding ItemName}" 
									ClickCommand="{Binding ItemSelectedCommand}" 
									ActionListFactory="{Binding ActionListFactory, RelativeSource={RelativeSource FindAncestor, AncestorType=controls:ProjectItemTreeControl}}" 
									ProjectItem="{Binding}"/>
								<controls:DisplayInWidgetControl 
									ProjectItem="{Binding}" 
									TabManager="{Binding TabManager, RelativeSource={RelativeSource TemplatedParent}}"/>
							</Panel>
						</TreeDataTemplate>
					</TreeView.ItemTemplate>
				</TreeView>
			</ControlTemplate>
		</Setter>
	</ControlTheme>
```

The TabManager property in ProjectItemTreeControl is getting bound fine, but the TabManager property in DisplayInWidgetControl is being set to null. 

I can't find any documentation surrounding binding properties between controls, am I trying to do something outside the scope of TemplateBinding?

**Comment by AdamIzon** [comment]
I'll note that changing the binding expression to `{Binding CurrentTab, RelativeSource={RelativeSource FindAncestor, AncestorType=controls:ProjectItemTreeControl}}` fixes my problem, but I would still like to know why it doesn't working with TemplateBinding

**Reply by guldud** [reply]
I use TemplatedControl in a TemplatedControl, and TemplateBinding for the Inner control does not work either.
If I remove outer control or simply copy the XAML from this control into the inner control, TemplateBinding works.

**Reply by timunie** [reply]
Yeah TemplateBinding only works inside ControlTemplates. They are very special. Your ItemTemplate, even if set inside ControlTemplate, is technically not part of it and thus fials to resolve. This is by design.

**Reply by guldud** [reply]
I see. Is there more to the story? Is there a way to combine templated controls and still use TemplateBinding? For example by using ResourceDictionary instead of Styles ? Or using ordinary binding with expressions like "$parent[TextBox].RevealPassword" ? The built-in controls is using each other a lot in the Simple and Fluent themes. And I'm trying to decipher the do's and don'ts. And how this "by design" is actually architected.

**Reply by AdamIzon** [reply]
> Yeah TemplateBinding only works inside ControlTemplates. They are very special. Your ItemTemplate, even if set inside ControlTemplate, is technically not part of it and thus fials to resolve. This is by design.

So the problem is that I'm TemplateBinding within an ItemTemplate in particular? Or is the problem that I'm TemplateBinding to a target that is nested inside another object?

Edit:
I'm starting to think it might be because I'm TemplateBinding within a DataTemplate/TreeDataTempalte. I guess I would have the same issue if I were to try to do a TemplateBinding from within any kind of DataTemplate?

**Reply by timunie** [reply]
https://docs.avaloniaui.net/docs/0.10.x/data-binding/binding-in-a-control-template 

In addtion: It must be all reachable in from ControlTemplate itself, so DataTemplates won't work.

**Reply by AdamIzon** [reply]
> https://docs.avaloniaui.net/docs/0.10.x/data-binding/binding-in-a-control-template
> 
> In addtion: It must be all reachable in from ControlTemplate itself, so DataTemplates won't work.

I don't see this kind of explanation in the docs you linked me. Perhaps it would be a good idea to update this page to explain template binding in more depth in the future? In fact I have found the documentation on binding in general to be quite shallow, so a new user has to look at a lot of external documentation if they want to go outside the narrow path the official documentation sets them up for.

**Reply by timunie** [reply]
yeah contributions to the docs are welcome, so if you think something was missing there feel free to provide a PR with the added or corrected info.
